File: BlocksWorld/BlocksWorld.pl
solve_blocks(Start, Goal, Solution) :-
    dfs(Start, Goal, [], Solution, 0, 10).

dfs(State, Goal, Path, Path, _, _) :-
    same_state(State, Goal).

dfs(State, Goal, CurrentPath, FinalPath, CurrentDepth, MaxDepth) :-
    CurrentDepth < MaxDepth,
    make_move(State, NextState, Move),
    \+ member(NextState, [State]),
    NewDepth is CurrentDepth + 1,
    dfs(NextState, Goal, [Move|CurrentPath], FinalPath, NewDepth, MaxDepth).

same_state([], []).
same_state(State1, State2) :-
    sort(State1, S1),
    sort(State2, S2),
    S1 = S2.

make_move(State, NewState, move(Block, table)) :-
    member(on(Block, Support), State),
    Support \= table,
    is_clear(Block, State),
    replace_position(State, on(Block, Support), on(Block, table), NewState).

make_move(State, NewState, move(Block, Target)) :-
    member(on(Block, Support), State),
    member(on(Target, _), State),
    is_clear(Block, State),
    is_clear(Target, State),
    Block \= Target,
    replace_position(State, on(Block, Support), on(Block, Target), NewState).

is_clear(Block, State) :-
    \+ member(on(_, Block), State).

replace_position([], _, _, []).
replace_position([Old|Rest], Old, New, [New|Rest]).
replace_position([Other|Rest], Old, New, [Other|NewRest]) :-
    Other \= Old,
    replace_position(Rest, Old, New, NewRest).

print_solution([]).
print_solution([move(Block, table)|Rest]) :-
    format('Move ~w to table~n', [Block]),
    print_solution(Rest).
print_solution([move(Block, Target)|Rest]) :-
    Target \= table,
    format('Move ~w onto ~w~n', [Block, Target]),
    print_solution(Rest).

example1 :-
    write('Initial: B on A, A on table'), nl,
    write('Goal: A on B, B on table'), nl, nl,
    
    Initial = [on(a, table), on(b, a)],
    Goal = [on(b, table), on(a, b)],
    
    (solve_blocks(Initial, Goal, Moves) ->
        (write('SOLUTION FOUND:'), nl,
         reverse(Moves, OrderedMoves),
         print_solution(OrderedMoves))
    ;
        write('No solution found')
    ), nl.


example2 :-
    write('Initial: A, B, C all on table'), nl,
    write('Goal: A on B, B on C, C on A (impossible circle)'), nl, nl,
    
    Initial = [on(a, table), on(b, table), on(c, table)],
    Goal = [on(a, b), on(b, c), on(c, a)],
    
    (solve_blocks(Initial, Goal, Moves) ->
        (write('SOLUTION FOUND:'), nl,
         reverse(Moves, OrderedMoves),
         print_solution(OrderedMoves))
    ;
        write('NO SOLUTION POSSIBLE')
    ), nl.

run_examples :-
    example1,
    example2,
    write('Program completed.'), nl.

:- initialization(run_examples).

File: BlocksWorld/BlocksWorld.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>
#include <algorithm>
using namespace std;

struct State {
    vector<pair<char, string>> on;
    vector<string> moves;
};

bool isGoal(const State &s, const vector<pair<char, string>> &goal) {
    for (auto &g : goal) {
        char block = g.first;
        string target = g.second;
        bool found = false;
        for (auto &p : s.on) {
            if (p.first == block && p.second == target) {
                found = true;
                break;
            }
        }
        if (!found) return false;
    }
    return true;
}


vector<char> getClearBlocks(const State &s) {
    vector<char> allBlocks;
    vector<char> hasSomethingOnTop;
    for (auto &p : s.on) {
        allBlocks.push_back(p.first);
        if (p.second != "Table") {
            hasSomethingOnTop.push_back(p.second[0]);
        }
    }
    vector<char> clear;
    for (char b : allBlocks) {
        if (find(hasSomethingOnTop.begin(), hasSomethingOnTop.end(), b) == hasSomethingOnTop.end()) {
            clear.push_back(b);
        }
    }
    return clear;
}

vector<State> Children(const State &current) {
    vector<State> nextStates;
    vector<char> clear = getClearBlocks(current);
    for (char x : clear) {
        string under;
        for (auto &p : current.on) {
            if (p.first == x) {
                under = p.second;
                break;
            }
        }
        if (under != "Table") {
            State newState = current;
            for (auto &p : newState.on) {
                if (p.first == x) {
                    p.second = "Table";
                    break;
                }
            }
            string move = "Move " + string(1, x) + " to Table";
            newState.moves.push_back(move);
            nextStates.push_back(newState);
        }
        for (char y : clear) {
            if (y == x) continue;
            State newState = current;
            for (auto &p : newState.on) {
                if (p.first == x) {
                    p.second = string(1, y);
                    break;
                }
            }
            string move = "Move " + string(1, x) + " onto " + string(1, y);
            newState.moves.push_back(move);
            nextStates.push_back(newState);
        }
    }
    return nextStates;
}

bool dfs(State start, const vector<pair<char, string>> &goal, int depthLimit) {
    stack<State> st;
    st.push(start);
    while (!st.empty()) {
        State current = st.top();
        st.pop();
        if (isGoal(current, goal)) {
            cout << "Goal reached!\nSequence of moves:\n";
            for (auto &m : current.moves) {
                cout << m << endl;
            }
            return true;
        }
        if ((int)current.moves.size() < depthLimit) {
            vector<State> nextStates = Children(current);
            for (auto &ns : nextStates) {
                st.push(ns);
            }
        }
    }
    return false;
}

int main() {
    int n;
    cout << "Enter number of blocks: ";
    cin >> n;
    State start;
    vector<pair<char, string>> goal;
    cout << "\nEnter start state (format: Block On):\n";
    for (int i = 0; i < n; i++) {
        char block;
        string on;
        cin >> block >> on;
        start.on.push_back({block, on});
    }
    cout << "\nEnter goal state (format: Block On):\n";
    for (int i = 0; i < n; i++) {
        char block;
        string on;
        cin >> block >> on;
        goal.push_back({block, on});
    }
    int depthLimit;
    cout << "\nEnter depth limit: ";
    cin >> depthLimit;
    if (!dfs(start, goal, depthLimit)) {
        cout << "Goal not found within depth limit.\n";
    }
    return 0;
}
